tests/recognizer-tests.js

```diff
@@ -129,6 +129,16 @@ test("A dynamic route recognizes", function() {
   equal(router.recognize("/zoo/baz"), null);
 });
 
+test("A dynamic route with encoded params", function() {
+  var handler = {};
+  var router = new RouteRecognizer();
+  router.add([{ path: "/foo/:bar", handler: handler }]);
+
+  equal(router.recognize("/foo/bar%20baz")[0].params.bar, "bar baz");
+  equal(router.recognize("/foo/bar%23baz")[0].params.bar, "bar#baz");
+  equal(router.recognize("/foo/bar%3fbaz")[0].params.bar, "bar?baz");
+});
+
```

Comment:
> the parsing of any route-recognizer DSLs like optional segments needs to happen before the URL is decoded.

Agreed wholeheartedly. You wouldn't want `/%2arest` to be interpreted as `/*rest`.

But what I'm saying is that currently no part of `path` ever gets `decodeURI` or `decodeURIComponent` called on it. That's a breaking API change.